prov/efa/src/rxr/rxr_pkt_type_req.c

```diff
@@ -823,46 +789,178 @@ ssize_t rxr_pkt_proc_matched_read_rtm(struct rxr_ep *ep,
+	return 0;
+}
+
 ssize_t rxr_pkt_proc_matched_medium_rtm(struct rxr_ep *ep,
 					struct rxr_rx_entry *rx_entry,
 					struct rxr_pkt_entry *pkt_entry)
 {
 	struct rxr_pkt_entry *cur;
+	struct rxr_pkt_entry *nxt;
 	struct efa_mr *desc;
 	char *data;
 	size_t offset, hdr_size, data_size;
+	int err;
 
 	cur = pkt_entry;
 	while (cur) {
-		hdr_size = rxr_pkt_req_hdr_size(pkt_entry);
+		cur->x_entry = rx_entry;
+		hdr_size = rxr_pkt_req_hdr_size(cur);
 		data = (char *)cur->pkt + hdr_size;
 		offset = rxr_get_medium_rtm_base_hdr(cur->pkt)->offset;
 		data_size = cur->pkt_size - hdr_size;
 		desc = rx_entry->desc[0];
-		ofi_copy_to_hmem_iov(desc ? desc->peer.iface : FI_HMEM_SYSTEM,
-				     desc ? desc->peer.device.reserved : 0,
-				     rx_entry->iov,
-				     rx_entry->iov_count,
-				     offset,
-				     data,
-				     data_size);
-		rx_entry->bytes_done += data_size;
-		cur = cur->next;
-	}
 
-	if (rx_entry->total_len == rx_entry->bytes_done) {
-		rxr_pkt_rx_map_remove(ep, pkt_entry, rx_entry);
-		/*
-		 * rxr_cq_handle_rx_completion() releases pkt_entry, thus
-		 * we do not release it here.
+		/* 
+		 * cur will be released by rxr_pkt_handle_rtm_data_copied(), so we have to
+		 * save cur->next here. we also set cur->next to NULL, which is required by
+		 * rxr_pkt_entry_release_rx().
 		 */
-		rxr_cq_handle_rx_completion(ep, pkt_entry, rx_entry);
-		rxr_msg_multi_recv_free_posted_entry(ep, rx_entry);
-		rxr_release_rx_entry(ep, rx_entry);
-		return 0;
+		nxt = cur->next; 
+		cur->next = NULL;
+
+		if (rxr_pkt_use_read_copy(desc, data_size)) {
+			err = rxr_pkt_post_read_to_copy_data(ep, cur, data, data_size, rx_entry, offset);
+			if (OFI_UNLIKELY(err))
+				return err;
+		} else {
+			err = ofi_copy_to_hmem_iov(desc ? desc->peer.iface : FI_HMEM_SYSTEM,
+						   desc ? desc->peer.device.reserved : 0,
+						   rx_entry->iov, rx_entry->iov_count,
+						   offset, data, data_size);
+			if (err < 0)
+				return err;
+			rxr_pkt_handle_medium_rtm_data_copied(ep, cur);
+		}
+		cur = nxt;
+	}
+
+	return 0;
+}
+
+static inline
+size_t rxr_pkt_proc_matched_long_rtm(struct rxr_ep *ep,
+				     struct rxr_rx_entry *rx_entry,
+				     struct rxr_pkt_entry *pkt_entry)
+{
+	size_t hdr_size, data_size;
+	struct efa_mr *desc;
+	char *data;
+	int err;
+
+	hdr_size = rxr_pkt_req_hdr_size(pkt_entry);
+	data = (char *)pkt_entry->pkt + hdr_size;
+	data_size = pkt_entry->pkt_size - hdr_size;
+	assert(data_size < rx_entry->total_len);
+
+	desc = rx_entry->desc[0];
+	if (rxr_pkt_use_read_copy(desc, data_size)) {
+		err = rxr_pkt_post_read_to_copy_data(ep, pkt_entry, data, data_size, rx_entry, 0);
+		if (err) {
+			rxr_cq_handle_rx_error(ep, rx_entry, err);
+			rxr_pkt_entry_release_rx(ep, pkt_entry);
+			return err;
+		}
+	} else {
+		err = ofi_copy_to_hmem_iov(desc ? desc->peer.iface : FI_HMEM_SYSTEM,
+					   desc ? desc->peer.device.reserved : 0,
+					   rx_entry->iov, rx_entry->iov_count, 0, data, data_size);
+		if (err < 0)
+			return err;
```

Comment:
Not sure I follow why this error block exits but the earlier one does some cleanup?